packages/vite/src/node/build.ts

```diff
@@ -875,13 +889,66 @@ async function buildEnvironment(
       prepareOutDir(resolvedOutDirs, emptyOutDir, environment)
     }
 
-    const res: RolldownOutput[] = []
-    for (const output of normalizedOutputs) {
-      res.push(await bundle[options.write ? 'write' : 'generate'](output))
-    }
+    const res = await build()
+
     logger.info(
       `${colors.green(`✓ built in ${displayTime(Date.now() - startTime)}`)}`,
     )
+
+    async function build() {
+      const res: RolldownOutput[] = []
+      for (const output of normalizedOutputs) {
+        // TODO(underfin): using the generate at development build could be improve performance.
+        res.push(await bundle![options.write ? 'write' : 'generate'](output))
+      }
+
+      if (server) {
+        // watching the files
+        for (const file of bundle!.watchFiles) {
+          if (path.isAbsolute(file) && fs.existsSync(file)) {
+            server.watcher.add(file)
+          }
+        }
+
+        // Write the output files to memory
+        for (const output of res) {
+          for (const outputFile of output.output) {
+            server.memoryFiles[outputFile.fileName] =
+              outputFile.type === 'chunk' ? outputFile.code : outputFile.source
+          }
+        }
+      }
+      return res
+    }
+
+    if (server) {
+      server.watcher.on('change', async (file) => {
+        const startTime = Date.now()
+        const hmrOutput = (await bundle!.generateHmrPatch([file]))!
+        if (hmrOutput.patch) {
+          const url = `${Date.now()}.js`
+          server.memoryFiles[url] = hmrOutput.patch
+          server.ws.send({
+            type: 'update',
+            updates: hmrOutput.hmrBoundaries.map((boundary) => {
+              return {
+                type: 'js-update',
+                url,
+                path: boundary.boundary,
+                acceptedPath: boundary.acceptedVia,
+                timestamp: 0,
+              }
+            }),
+          })
+          logger.info(
+            `${colors.green(`✓ Found ${path.relative(root, file)} changed, rebuilt in ${displayTime(Date.now() - startTime)}`)}`,
+          )
+
+          await build()
```

Comment:
I guess a stale content would be shown if the dev server is accessed before this `await build()` is called. When it's accessed in that state, I guess we need to send the patch together with the bundle to the browser.

Reply:
We discuss it at rolldown team. It could be happened, but it is correct behavior. It just like the process at vite/webpack, the file changed -> vite hmr or webpack build process (the user visit page), the user also can't seen the updated content.

Reply:
> It just like the process at vite/webpack, the file changed -> vite hmr or webpack build process (the user visit page), the user also can't seen the updated content.

I'm not sure if I get this. But in Vite, I guess that never happened because Vite doesn't bundle the files.

Reply:
Assume vite hmr is working to invalidate modules at `propagateUpdate`, but it is not finished, the user visit page, the user could be see original content.

Reply:
In that case, won't the new module be seen by users? Because Vite doesn't use the HMR content when accessing the page.

Reply:
Yeah.